### ISSUE TITLE
EnterpriseGatewayApp.max_kernels_per_user not working when KERNEL_USERNAME not provided

### ISSUE THREAD

**Issue by amangarg96** [jupyter/enterprise_gateway#490]
I launched Jupyter Enterprise Gateway with the parameter `--EnterpriseGatewayApp.max_kernels_per_user=2` but it's not restricting the user (i.e each jupyter-lab server) to launching just 2 kernels. He is able to launch any number of kernels.

What could be the reason for this behaviour?

I am using Jupyter Enterprise Gateway to run iPython kernels in YARN Cluster Mode on Apache Spark.
The Jupyter Lab server is running on a local machine (Macbook), while the Jupyter Enterprise Gateway server is running on one of the Nodes of the cluster, and kernels are launched on the cluster.

**Comment by kevin-bates** [comment]
Hmm. This sounds like the command-line option is not getting recognized and remaining at the default of `-1` (which disables this functionality).

If I use a user name of 'alice' (i.e., I set env `KERNEL_USERNAME = alice` prior to launching the notebook/lab) and `--EnterpriseGatewayApp.max_kernels_per_user=2`, I get the following when attempting to start a third kernel (with the other 2 still running): 
```
    value = future.result()
  File "/opt/conda/lib/python3.6/site-packages/tornado/gen.py", line 1141, in run
    yielded = self.gen.throw(*exc_info)
  File "/opt/conda/lib/python3.6/site-packages/nb2kg/managers.py", line 64, in fetch_kg
    response = yield client.fetch(url, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/tornado/gen.py", line 1133, in run
    value = future.result()
tornado.httpclient.HTTPClientError: HTTP 403: A max kernels per user limit has been set to 2 and user 'alice' currently has 2 active kernels.
```
On the EG server, this appears as the following in its log:
```
[E 2018-11-14 07:36:54.771 EnterpriseGatewayApp] A max kernels per user limit has been set to 2 and user 'alice' currently has 2 active kernels.
[W 181114 07:36:54 web:2162] 403 POST /api/kernels (9.85.140.70) 21.61ms
```
Even if you are not explicitly setting `KERNEL_USERNAME`, EG will use the username of its process, which would behave as if an overall max-kernels limit was imposed.

Although this functionality has been present for a while, what version of EG are you running?

You might also instead try setting the environment variable option (`EG_MAX_KERNELS_PER_USER`).  If that works, then it's an issue with how you're setting the command line option.

If you have the repo pulled, you could also run the following...
```
nosetests -v enterprise_gateway.tests.test_enterprise_gateway:TestEnterpriseGateway.test_max_kernels_per_user
```
which should yield...
```
Number of kernels should be limited per user. ... ok

----------------------------------------------------------------------
Ran 1 test in 1.123s

OK
```

**Comment by amangarg96** [comment]
When I set env `KERNEL_USERNAME = alice` prior to launching the lab, the max_kernels_per_user restriction is working like you mentioned above, but it's not working on unsetting the `KERNEL_USERNAME` environment variable. 
Maybe EG is not able to identify that the kernels are coming from the same user, when the `KERNEL_USERNAME` is not set and sees them as new users. 

I'm using version 1.1.0 of jupyter-enterprise-gateway

**Comment by kevin-bates** [comment]
Thanks for the additional information.  I see the issue and should have a fix shortly.

The issue is that when `KERNEL_USERNAME` is not specified in the request, the kernel sessions are created under a username value of 'unspecified'.  However, at the time the active kernel limit is enforced, the server has set a `KERNEL_USERNAME` to the process' username.  Since this value doesn't match with `unspecified`, the limit is not enforced.  I suppose another workaround would be to run your EG process as the user `unspecified`. 😃   The fix will be to ensure those two areas of the code are coordinated.

In general, you are strongly advised to set `KERNEL_USERNAME` on the client.  When that is not the case, your application will not be able to leverage the full capabilities of EG.  Things like the [authorized/unauthorized users lists or impersonation](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/getting-started-security.html#security-features) will not function as intended.

**Comment by kevin-bates** [comment]
@amangarg96 - I've also updated the 1.x with these changes.  You should be able to build from there.  We'll be performing some release cycles in the next couple weeks.

**Comment by amangarg96** [comment]
Thank you for promptly fixing the bug :)

**Comment by kevin-bates** [comment]
@amangarg96 - FYI - we have included this change in our [1.1.1](https://github.com/jupyter/enterprise_gateway/releases/tag/v1.1.1) release.
